modeling/local-slice-algorithm.go

```diff
@@ -0,0 +1,132 @@
+
+import (
+	"container/list"
+	"fmt"
+	"math"
+)
+
+// LocalSliceAlgorithm 'borrows' and 'rents' endpoints from other zones to make
+// the local EndpointSlice balanced with the incoming traffic (number of nodes
+// distribution):
+// 1. The proportion of incoming traffic for a zone is calculated based on the
+// proportion of nodes or cores in the zone.
+// 2. This is compared with the proportion of endpoints in the zone to calculate
+// the deviation from the expected number of endpoints in a perfectly balanced
+// system.
+// 3. EndpointSlices in zones with less endpoints than expected will receive
+// endpoints from zones that have more endpoints than expected.
+type LocalSliceAlgorithm struct{}
+
+type endpointDeviation struct {
+	name      string
+	deviation int
+}
+
+// CreateSliceGroups creates sliceGroups with 'one local EndpointSliceGroup per
+// zone' policy
+func (alg LocalSliceAlgorithm) CreateSliceGroups(region regionInfo) (map[string]EndpointSliceGroup, error) {
+	if region.zoneDetails == nil {
+		return nil, fmt.Errorf("zoneDetail should not be nil")
+	}
+	sliceGroups := map[string]EndpointSliceGroup{}
+	// endpointsAvailable stores 'endpointDeviation' -- zones with 'extra'
+	// endpoints
+	endpointsAvailable := list.New()
+	// endpointsNeeded stores zones with number of endpoints needed
+	endpointsNeeded := map[string]int{}
+	for zoneName, zone := range region.zoneDetails {
+		var localGroup EndpointSliceGroup
+		localGroup.Label = zoneName
+		// this local sliceGroup should only receive traffic from current zone,
+		// this map tracks weights of traffic from different zones to this
+		// sliceGroup
+		localGroup.Weights = map[string]float64{zoneName: 1.0}
+		// this map stores composition of this sliceGroup, zoneName - number of
+		// endpoints from zoneName
+		localGroup.Composition = map[string]int{}
+
+		// the expected number of endpoints is with the same ratio as the number
+		// of nodes in this zone
+		expectedEndpoints := zone.nodesRatio * float64(region.totalEndpoints)
+		// deviation: - means need more endpoints from other sliceGroups, +
+		// means need give out endpoints to other sliceGroups
+		deviation := math.Round(float64(zone.Endpoints) - expectedEndpoints)
+		if deviation > 0 {
+			endpointsAvailable.PushBack(endpointDeviation{name: zoneName, deviation: int(deviation)})
+			localGroup.Composition[zoneName] = int(expectedEndpoints)
+		} else {
+			endpointsNeeded[zoneName] = int(-deviation)
+			localGroup.Composition[zoneName] = zone.Endpoints
+		}
+
+		sliceGroups[zoneName] = localGroup
+	}
+
+	err := balanceSliceGroups(endpointsAvailable, endpointsNeeded, sliceGroups)
+	if err != nil {
+		return nil, err
+	}
+	return sliceGroups, nil
+}
+
+// balanceSliceGroups helps to 'rent' endpoints from zones have extra endpoints
+// to EndpointSliceGroups that have insufficient ones
+func balanceSliceGroups(endpointsAvailable *list.List, endpointsNeeded map[string]int, sliceGroups map[string]EndpointSliceGroup) error {
```

Comment:
This function still feels pretty complex. Maybe it would be possible to restructure this a bit further? I think the balancing could benefit from some kind of struct like this. This is pretty rough and probably doesn't work as is, but hopefully it provides an idea of how it could be structured. Something like this would be a great place to add some unit tests.

```golang
type endpointsAvailable struct {
  byZone map[string]int
}

func (ea *endpointsAvailable) takeEndpoints(n int) map[string]int {
  takenByZone := map[string]int{}
  for zone, num := range ea.byZone {
    if num >= n {
      takenByZone[zone] = n
      ea.byZone[zone] = num - n
      break
    } else {
      takenByZone[zone] = num
      ea.byZone[zone] = 0
      n -= num
    }
  }
  return takenByZone
}
```